doc: release-notes: Add 3.5.0 release notes for networking (#63703)
zephyrproject-rtos/zephyr: doc/releases/release-notes-3.5.rst

  * Added new API functions:

    * :c:func:`coap_has_descriptive_block_option`
    * :c:func:`coap_remove_descriptive_block_option`
    * :c:func:`coap_packet_remove_option`
    * :c:func:`coap_packet_set_path`

* Connection Manager:

  * Added support for auto-connect and auto-down behaviors (controlled by
    :c:enum:`CONN_MGR_IF_NO_AUTO_CONNECT` and :c:enum:`CONN_MGR_IF_NO_AUTO_DOWN`
    flags).
  * Split Connection Manager APIs into separate header files.
  * Extended Connection Manager documentation to cover new functionalities.

* DHCP:

  * Added support for DHCPv4 unicast replies processing.
  * Added support for DHCPv6 protocol.

* ICMP:

  * Fixed ICMPv6 error message type check.
  * Reworked ICMP callback registration and handling, which allows to register
    multiple handlers for the same ICMP message.
  * Introduced an API to send ICMP Echo Request (ping).
  * Added possibility to register offloaded ICMP ping handlers.
  * Added support for setting packet priority for ping.

* IPv6:

  * Made sure that ongoing DAD procedure is cancelled when IPv6 address is removed.
  * Fixed a bug, where Solicited-Node multicast address could be removed while
    still in use.

> * LwM2M:

Reviewer: btw I would add a mention to the new `lwm2m create` command - it's a great tool and worth advertising :)

Author: True, I've mentally linked it to IoP testing and skipped a separate entry for it, fixed now.

  * Added support for tickless mode. This removes the 500 ms timeout from the socket loop
    so the engine does not constantly wake up the CPU. This can be enabled by
    :kconfig:option:`CONFIG_LWM2M_TICKLESS`.
  * Added new :c:macro:`LWM2M_RD_CLIENT_EVENT_DEREGISTER` event.
  * Block-wise sending now supports LwM2M read and composite-read operations as well.

>     When :kconfig:option:`LWM2M_COAP_BLOCK_TRANSFER` is enabled, any content that is larger than

Reviewer: ```suggestion
    When :kconfig:option:`CONFIG_LWM2M_COAP_BLOCK_TRANSFER` is enabled, any content that is larger than
```

  * Block-wise transfers don't require tokens to match anymore as this was not in line
    with CoAP specification (CoAP doesn't require tokens re-use).
  * Various fixes to bootstrap. Now client ensures that Bootstrap-Finnish command is send,
    before closing the DTLS pipe. Also allows Bootstrap server to close the DTLS pipe.
    Added timeout when waiting for bootstrap commands.
  * Added support for X509 certificates.
  * Various fixes to string handling. Allow setting string to zero length.
    Ensure string termination when using string operations on opaque resources.
  * Added support for Connection Monitoring object version 1.3.
  * Added protection for Security object to prevent read/writes by the server.
  * Fixed a possible notification stall in case of observation token change.
  * Added LwM2M interoperability test-suite against Leshan server.
  * API documentation improvements.
  * Several other minor fixes and improvements.

* Misc:

  * Time and timestamps in the network subsystem, PTP and IEEE 802.15.4
    were more precisely specified and all in-tree call sites updated accordingly.
    Fields for timed TX and TX/RX timestamps have been consolidated. See

>     :c:type:`net_time_t`, :c:struct:`net_ptp_time`, :c:struct:`ieee802154_config`,

Reviewer: `net_time_t` does not link to anything as `net_time.h` does not define any doxygen group nor does it appear in the Sphinx documentation. May be worth fixing?

Author: TBH I kind of miss an idea where to put such a reference (and least or now) and defining a whole new doc page for a header with a single typedef sounds like a overshot to me.

Reviewer: > TBH I kind of miss an idea where to put such a reference (and least or now) and defining a whole new doc page for a header with a single typedef sounds like a overshot to me.

Ya, I feel you. Would it make sense to start creating some kind of "Utilities" group under Connectivity? https://docs.zephyrproject.org/latest/doxygen/html/group__connectivity.html, and then maybe an RST as a subpage of https://docs.zephyrproject.org/latest/connectivity/networking/api/apis.html.

In case this helps identify other candidates for maybe grouping other APIs alongside `net_time_t` (socketutils?). Note that these are also currently not showing in the Sphinx docs so you might want to have a look in any case.

```
❯ grep -L -r '@defgroup\|@addtogroup' include/zephyr/net/
include/zephyr/net/http/parser_url.h
include/zephyr/net/http/parser.h
include/zephyr/net/http/parser_state.h
include/zephyr/net/http/service.h
include/zephyr/net/loopback.h
include/zephyr/net/socket_offload.h
include/zephyr/net/ieee802154_pkt.h
include/zephyr/net/ipv4_autoconf.h
include/zephyr/net/net_time.h
include/zephyr/net/socketutils.h
include/zephyr/net/ieee802154_radio_openthread.h
include/zephyr/net/bt.h
include/zephyr/net/canbus.h
```

Author: For now I've just added a new section under "Time Sensitive Networking" for net_time. Other headers likely need a review, but from what I see in many cases they either not use doxygen at all, or are documenting internal APIs.

    documentation. As this is largely an internal API, existing applications will
    most probably continue to work unchanged.
  * Added support for additional net_pkt filter hooks:

    * :kconfig:option:`CONFIG_NET_PKT_FILTER_IPV4_HOOK`
    * :kconfig:option:`CONFIG_NET_PKT_FILTER_IPV6_HOOK`
    * :kconfig:option:`CONFIG_NET_PKT_FILTER_LOCAL_IN_HOOK`

  * Reworked several networking components to use timepoint API.
  * Added API functions facilitate going through all IPv4/IPv6 registered on an
    interface (:c:func:`net_if_ipv4_addr_foreach`, :c:func:`net_if_ipv6_addr_foreach`).
  * ``NET_EVENT_IPV6_PREFIX_ADD`` and ``NET_EVENT_IPV6_PREFIX_DEL`` events now provide
    more detailed information about the prefix (:c:struct:`net_event_ipv6_prefix`).
  * General cleanup of the shadowed variables across the networking subsystem.
  * Added ``qemu_cortex_a53`` networking support.
  * Introduced new modem subsystem.
  * Added new :zephyr:code-sample:`cellular-modem` sample.
  * Added support for network interface names (instead of reusing underlying device name).
  * Removed support for Google Cloud IoT sample due to service retirement.

>   * Fixed a bug where packets passed in promiscuous mode could've been modified

Reviewer: nit 

```suggestion
  * Fixed a bug where packets passed in promiscuous mode could have been modified
```

  * Prevent PPP connection reestablish when carrier is down.

* Sockets:

  * Added support for statically allocated socketpairs (in case no heap is available).
  * Made send timeout configurable (:kconfig:option:`CONFIG_NET_SOCKET_MAX_SEND_WAIT`).
  * Added support for ``FIONREAD`` and ``FIONBIO`` :c:func:`ioctl` commands.
  * Fixed input filtering for connected datagram sockets.
  * Fixed :c:func:`getsockname` operation on unconnected sockets.
  * Added new secure socket options for DTLS Connection ID support:

    * :c:macro:`TLS_DTLS_CID`
    * :c:macro:`TLS_DTLS_CID_VALUE`
    * :c:macro:`TLS_DTLS_PEER_CID_VALUE`
    * :c:macro:`TLS_DTLS_CID_STATUS`

  * Added support for :c:macro:`SO_REUSEADDR` and :c:macro:`SO_REUSEPORT` socket options.

* TCP:

  * Fixed potential stall in data retransmission, when data was only partially acknowledged.
  * Made TCP work queue priority configurable (:kconfig:option:`CONFIG_NET_TCP_WORKER_PRIO`).
  * Added support for TCP new Reno collision avoidance algorithm.
  * Fixed source address selection on bound sockets.
  * Fixed possible memory leak in case listening socket was closed during active handshake.

>   * Fixed RST packet handing during handshake.

Reviewer: ```suggestion
  * Fixed RST packet handling during handshake.
```
